mapie/mondrian.py

```diff
@@ -0,0 +1,419 @@
+        """
+
+        X, y, groups = self._check_fit_parameters(X, y, groups)
+        self._check_group_length(X, groups)
+        self.unique_groups = np.unique(groups)
+        self.mapie_estimators = {}
+        if isinstance(self.mapie_estimator, MapieClassifier):
+            self.n_classes = len(np.unique(y))
+
+        for group in self.unique_groups:
+            mapie_group_estimator = deepcopy(self.mapie_estimator)
+            indices_groups = np.argwhere(groups == group)[:, 0]
+            X_g, y_g = X[indices_groups], y[indices_groups]
+            mapie_group_estimator.fit(X_g, y_g, **fit_params)
+            self.mapie_estimators[group] = mapie_group_estimator
+        return self
+
+    def predict(
+        self,
+        X: ArrayLike,
+        groups: ArrayLike,
+        alpha: Optional[Union[float, Iterable[float]]] = None,
+        **predict_params
+    ) -> Union[NDArray, Tuple[NDArray, NDArray]]:
+        """
+        Perform conformal prediction for each group of individuals
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            The input data
+
+        groups : ArrayLike of shape (n_samples,)
+            The groups of individuals. Must be defined by integers.
+
+        alpha : float or Iterable[float], optional
+            The desired coverage level(s) for each group.
+
+            By default None.
+
+        **predict_params
+            Additional keyword arguments to pass to the estimator's predict
+            method that may be specific to the Mapie estimator used
+
+        Returns
+        -------
+        y_pred : NDArray of shape (n_samples,) or (n_samples, n_outputs)
+            The predicted values
+
+        y_pss : NDArray of shape (n_samples, n_outputs, n_alpha)
+            The predicted sets for the desired levels of coverage
+        """
+
+        check_is_fitted(self, self.fit_attributes)
+        X = cast(NDArray, X)
+        groups = self._check_groups_predict(X, groups)
+        alpha_np = cast(NDArray, check_alpha(alpha))
+        if alpha_np is None and self.mapie_estimator.estimator is not None:
+            return self.mapie_estimator.estimator.predict(
+                X, **predict_params
+            )
+        else:
+            if isinstance(self.mapie_estimator, MapieClassifier):
+                y_pred = np.empty(
+                    (X.shape[0], )
+                )
+                y_pss = np.empty(
+                    (X.shape[0], self.n_classes, len(alpha_np))
+                )
+            else:
+                y_pred = np.empty((X.shape[0],))
+                y_pss = np.empty((X.shape[0], 2, len(alpha_np)))
+            unique_groups = np.unique(groups)
+            for i, group in enumerate(unique_groups):
+                indices_groups = np.argwhere(groups == group)[:, 0]
+                X_g = X[indices_groups]
+                y_pred_g, y_pss_g = self.mapie_estimators[group].predict(
```

Comment:
I might have missed this, but where do we check if a new group appears in the `predict` (that was not in the `fit`)?

Reply:
yes, chekced in `_check_partition_predict`